src/daft-core/src/array/from_iter.rs

```diff
@@ -40,6 +41,23 @@ where
         let data_array: PrimitiveArray<_> = array.into();
         Self::new(field, data_array.boxed()).unwrap()
     }
+
+    pub fn from_regular_iter<F, I>(field: F, iter: I) -> DaftResult<Self>
```

Comment:
Unfortunately, `from_iter` is already implemented just a few lines above. The difference is that *that* `from_iter` actually requires that `I: TrustedLen`, whereas this new one has a lighter restriction in that `I: Iterator` only.

I would actually like to change the old `from_iter` -> `from_trusted_len_iter` and this new function from `from_regular_iter` -> `from_iter`. However, I wanted to not introduce too much confusion in this PR. Perhaps a separate PR to clean up these names?

Reply:
From my read-throughs of the `daft-core` crate thus far, I believe that our story around `DataArray` creation is quite unintuitive. I constantly struggle to remember which API to use. Most of the available APIs also require the additional restriction that `I: TrustedLen`. Looking at the source code for `TrustedLen`, not many iterators implement it. It's quite rough to work with.

Until `TrustedLen` stabilizes, I guess we have to use the one provided by `arrow2`.

I think it would be quite helpful to clean up the story around `DataArray`.

Reply:
> I would actually like to change the old from_iter -> from_trusted_len_iter and this new function from from_regular_iter -> from_iter. However, I wanted to not muddy this PR with adjacent changes. Perhaps a separate PR to clean up these names?

👍 I agree with this. We make an issue so this doesn't get lost.

Reply:
You could technically created `TrustedLen`s actually if you construct an [`IndexRange`](https://docs.rs/arrow2/latest/arrow2/types/struct.IndexRange.html) and map on that in your `list_unique_count` implementations. It's not the most elegant, but just an idea